NavigationPolygon: Implement `get`/`set_polygon` fast paths.

libklein opened the pull request:
Follow up to PR #92955. This PR implements the changes outlined in https://github.com/godotengine/godot/pull/92955#discussion_r1635665360. 

As a first step, it implements fast-path `get/set_polygons` and `get/set_outlines` functions, updating `NavMeshGenerator2D` accordingly. Implemented `get/set_polygons` for `NavigationMesh` as well, replacing `add_polygon` whenever a direct update was preferable.

libklein committed:
NavigationPolygon: Implement get/set_polygon fast paths.

AThousandShips commented on scene/resources/2d/navigation_polygon.h:
```diff
@@ -108,14 +105,16 @@ class NavigationPolygon : public Resource {
 	void remove_outline(int p_idx);
 	int get_outline_count() const;
+	void set_outlines(const Vector<Vector<Vector2>> &p_outlines);
+	const Vector<Vector<Vector2>> get_outlines() const;
```
```suggestion
	Vector<Vector<Vector2>> get_outlines() const;
```
Unnecessary

AThousandShips commented on modules/navigation/3d/nav_mesh_generator_3d.cpp:
```diff
@@ -933,9 +934,10 @@ void NavMeshGenerator3D::generator_bake_from_source_geometry_data(Ref<Navigation
 			nav_indices.write[2] = recast_index_to_native_index[index3];
 
-			p_navigation_mesh->add_polygon(nav_indices);
+            nav_polygons.push_back(nav_indices);
```
```suggestion
			nav_polygons.push_back(nav_indices);
```
Wrong indentation type

smix8 commented:
Needs both a rebase for 4.4 and the newer thread locks added to the functions.

libklein commented:
Rebased onto the latest master:
* added read/write locks where appropriate

smix8 commented on scene/resources/3d/importer_mesh.cpp:
```diff
@@ -1098,15 +1102,7 @@ Ref<NavigationMesh> ImporterMesh::create_navigation_mesh() {
-		v3.write[2] = face_indices[i + 2];
-		nm->add_polygon(v3);
-	}
+	nm->set_polygons(face_polygons);
```
While not focus of this PR this line and the above set_vertices() should be combined and use the newer set_data() to keep both vertices and polygons arrays thread-safe and in sync.

smix8 commented on modules/navigation/3d/nav_mesh_generator_3d.cpp:
```diff
@@ -939,6 +939,7 @@ void NavMeshGenerator3D::generator_bake_from_source_geometry_data(Ref<Navigation
 			nav_polygons.push_back(nav_indices);
 		}
 	}
+	p_navigation_mesh->set_polygons(nav_polygons);
```
This line is not needed as the newer set_data() is used below to set the vertices and polygons arrays in a thread-safe way so they stay in sync.

libklein commented:
I've added two commits. https://github.com/godotengine/godot/pull/93171/commits/d121458cbbee0055db1ab3824e15e30a74b24080 addresses the issues you've outlined.

https://github.com/godotengine/godot/pull/93171/commits/53d5b05f0456a50845cf027b4f4b17fbe0eb67b2 adds a overload to set_data that allows to set vertices, polygons, and outlines simultaneously in a thread safe way. Moreover, it avoids a for loop in get/set_data. Happy to push that to a different PR/revert if required.

smix8 commented:
Looks fine. I dont have any objection against the overload, that at least makes it clear that this function should never get exposed by accident.

akien-mga commented:
Looks good! Could you squash the commits? See [PR workflow](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase) for instructions.

akien-mga commented:
Something went wrong with your rebase, which seemed to have amended changes into an unrelated merge commit from the `master` branch:
![image](https://github.com/user-attachments/assets/4a75fb61-5659-4d5b-a104-935dc336519f)

I fixed it for you by resetting to the previous commit (53d5b05) and just running `git rebase -i upstream/master` (with upstream pointing at `godotengine/godot` and up to date), and marking commits 2 to 5 as `fixup`.

akien-mga commented:
Thanks!

libklein commented:
Thank you! Is there any other issue I could assist with?
